### PR TITLE
Test invoke contract dynamic

### PR DESCRIPTION
**Pull request by tcoratger**
## Checklist
- [ ] Linked to Github Issue
- [ ] Unit tests added
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.

### CODE REVIEW

File: src/executor/jit.rs [hunk 1]
```diff
@@ -154,3 +154,118 @@ impl<'m> JitNativeExecutor<'m> {
...
+            .invoke_contract_dynamic(
+                &function_id,
+                &call_data,
+                Some(initial_gas),
+                TestSyscallHandler,
+            )
+            .unwrap();
+
+        // Print the result of the contract invocation.
+        println!("result {:?}", result);
```
Comment (fmoletta) [inline src/executor/jit.rs:269]: Shouldn't this test check the value of result instead of just printing it?

**Review by fmoletta — CHANGES_REQUESTED** [review]

File: src/executor/jit.rs (hunk 1 above)
Comment (tcoratger) [reply src/executor/jit.rs:269]: Yes sure but here just it doesn't work. It's a PR to isolate the bug we have compiling Kakarot with Cairo native during the validate step. @igaray @edg-l can you confirm there is a bug here? The result in this print is 

```
ContractExecutionResult { remaining_gas: 340282366920938463463374607431768204045, failure_flag: true, return_values: [Felt(FieldElement { value: UnsignedInteger { limbs: [449000750542568112, 298359026771569608, 16599260449058738440, 8029189093433763828] } })], error_msg: Some("Failed to deserialize param #1") }
```

which seems not to be normal here

**Comment by azteca1998** [comment]
Hello, I tried running your contract in the main branch (not this PR, but copying everything this PR has) and in my machine the error was different:
```
ContractExecutionResult {
    remaining_gas: 340282366920938463463374607431768181385,
    failure_flag: true,
    return_values: [
        Felt(
            FieldElement {
                value: UnsignedInteger {
                    limbs: [
                        324175748156660941,
                        8931066404368655594,
                        15123115999449940179,
                        13175938098900479932,
                    ],
                },
            },
        ),
    ],
    error_msg: Some(
        "Input too long for arguments",
    ),
}
```

This just confirms my suspicion that the calldata is wrong for this specific contract. Since the calldata is hardcoded in the test let's analyze it:
```
To deserialize: `Array<Call>` (extracted from the Sierra program)
where `Call` is (extracted from the corelib):
```cairo
#[derive(Drop, Serde, Debug)]
pub struct Call {
    pub to: ContractAddress,
    pub selector: felt252,
    pub calldata: Span<felt252>
}
```

This means that the calldata is:
```
0x1     -- Number of `Call`s (array length).
0x1     -- #0's `Call::to` field.
0x7099f594eb65e00576e1b940a8a735f80bf7604ac401c48627045c4cc286f0
        -- #0's `Call::selector` field.
0x26    -- #0's `Call::calldata` field length (the next 38 elements are the data).
0x2
0xe4
0x84
0x4b
0x4b
0x52
0x54
0x80
0x80
0x80
0x83
0xf
0x42
0x40
0x94
0x0
0x0
0x0
0x0
0x0
0x0
0x0
0x0
0x0
0x0
0x0
0x0
0x0
0x0
0x0
0x0
0x0
0x0
0x0
0x0
0x12
0x34
0x80

0x80    -- Extra garbage.
0xc0    -- Extra garbage.
```

If we remove the extra garbage then the program returns the following:
```
ContractExecutionResult {
    remaining_gas: 340282366920938463463374607431767964335,
    failure_flag: true,
    return_values: [
        Felt(
            FieldElement {
                value: UnsignedInteger {
                    limbs: [
                        132997066441348640,
                        18446744073564198872,
                        13542802803994314020,
                        8112183345115268369,
                    ],
                },
            },
        ),
    ],
    error_msg: Some(
        "EOA: reentrant call",
    ),
}
```

All those errors are not native-related but are instead generated by the Cairo program itself. For example:
  - `Failed to deserialize param #1` is autogenerated by the Starknet compiler.
  - `Input too long for arguments` is also autogenerated by the Starknet compiler.
  - `EOA: reentrant call` I think it comes from [OpenZeppelin](https://docs.openzeppelin.com/contracts/2.x/api/utils#ReentrancyGuard).

**Comment by greged93** [comment]
Hey @azteca1998, we can retry this but what we were seeing on SiR was: calling __validate__ would return the `Failed to deserialize param #1`, however ONLY if we included the eth transaction validation code, sending the same calldata and removing the code would pass.

Also, the __execute__ entrypoint would always work, so setting `skip_validate` to true would just continue execution, even though AFAIK the args to __validate__ and __execute__ are the same.

Will try rerunning on SiR and keep you posted.

**Comment by azteca1998** [comment]
If what you sent me on Telegram is the same contract as in this issue then the reentrant call error may be thrown [here](https://github.com/kkrt-labs/kakarot-ssj/blob/main/crates/contracts/src/account_contract.cairo#L150). It's probably what you said.

As for the other line, I don't see how that line could cause the `Failed to deserialize param #1` error. If that line is causing the error it should crash with `failed to validate eth tx`. The only place where that error could be generated is [here](https://github.com/kkrt-labs/kakarot-ssj/blob/main/crates/utils/src/eth_transaction.cairo#L464-L471).

I'm not sure how easy or difficult it is to test that independently, but could you extract somehow the data involved in the statement that's failing? This way I may be able to test it better.

**Comment by greged93** [comment]
@azteca1998 I will try, will copy paste the code and try to repro without having the compiled contract

**Pull request closed without merging**